### PR TITLE
fix(security): sanitize request bodies against XSS #247

### PR DESCRIPTION
**Pull request by SmitRAmoliya**
Summary:
Add recursive server-side body sanitization for state-changing API requests so stored XSS payloads are stripped before controllers persist input.

Changes:
- install xss in the backend package
- add backend/src/middleware/sanitize-input.ts for recursive string sanitization across nested objects and arrays
- apply sanitizeRequestBody at the /api mount alongside CSRF protection
- add focused tests covering HTML stripping and GET passthrough behavior

Verification:
- backend/__tests__/input-sanitization.test.ts: 2/2 passing
- backend build still shows pre-existing generic db typing errors on develop; no new Step 6-specific compile issue found

Related Issues:
References #247

### CODE REVIEW

File: backend/src/middleware/sanitize-input.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import type { NextFunction, Request, Response } from 'express';
+import xss from 'xss';
+
+function sanitizeValue(value: unknown): unknown {
+  if (typeof value === 'string') {
+    return xss(value, {
+      whiteList: {},
+      stripIgnoreTag: true,
+      stripIgnoreTagBody: ['script', 'style'],
+    });
```
Comment (savitasawanteisg-byte) [inline backend/src/middleware/sanitize-input.ts]: **Bug / DoS Risk: No recursion depth limit**

The sanitizeValue function recurses into nested objects without any depth limit. An attacker could send a deeply nested JSON payload (10000+ levels) to trigger a stack overflow and crash the server.

**Suggested fix:** Add a depth parameter with a max limit (e.g. 20) and return the value unsanitized if exceeded.

File: backend/src/middleware/sanitize-input.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+  if (Array.isArray(value)) {
+    return value.map((item) => sanitizeValue(item));
+  }
+
+  if (value && typeof value === 'object') {
+    const entries = Object.entries(value as Record<string, unknown>).map(([key, nestedValue]) => [
+      key,
+      sanitizeValue(nestedValue),
+    ]);
```
Comment (savitasawanteisg-byte) [inline backend/src/middleware/sanitize-input.ts]: **Security: Prototype pollution risk**

Object.entries() will include __proto__ if present in the input, and Object.fromEntries() will recreate it. Explicit filtering of dangerous keys (__proto__, constructor, prototype) is recommended.

**Suggested fix:** Filter out __proto__, constructor, and prototype keys before calling Object.fromEntries().

File: backend/src/middleware/sanitize-input.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+  return value;
+}
+
+export function sanitizeRequestBody(req: Request, _res: Response, next: NextFunction): void {
+  if (['GET', 'HEAD', 'OPTIONS'].includes(req.method)) {
+    next();
+    return;
+  }
+
```
Comment (savitasawanteisg-byte) [inline backend/src/middleware/sanitize-input.ts:44]: **Security: Incomplete XSS protection - req.query and req.params not sanitized**

This middleware only sanitizes req.body. XSS payloads commonly arrive via query parameters (e.g. ?search=<script>alert(1)</script>). For comprehensive protection, query and params should also be sanitized.

Consider renaming the function to sanitizeRequest and also processing req.query.

File: backend/src/middleware/sanitize-input.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  req.body = sanitizeValue(req.body);
+  next();
+}
```
Comment (savitasawanteisg-byte) [inline backend/src/middleware/sanitize-input.ts]: **Code quality: Missing newline at end of file**

POSIX standard requires files to end with a newline. Add a trailing newline after the closing brace.

File: backend/__tests__/input-sanitization.test.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  });
+
+  it('does not interfere with safe GET requests', async () => {
+    const app = buildApp();
+    const response = await request(app).get('/view');
+
+    expect(response.status).toBe(200);
+    expect(response.body).toEqual({ ok: true });
+  });
+});
```
Comment (savitasawanteisg-byte) [inline backend/__tests__/input-sanitization.test.ts]: **Testing: Missing edge-case tests**

The test suite should include:
1. Deeply nested object - verify no stack overflow or depth is capped
2. Prototype pollution attempt - { __proto__: { isAdmin: true } } should not pollute
3. Non-string primitives - numbers, booleans, null should pass through unchanged
4. Empty/undefined body should not throw

Also missing newline at end of file.

File: src/__tests__/api/register.test.ts [hunk 6]
```diff
@@ -93,54 +109,57 @@ describe('POST /api/auth/register', () => {
     expect(stored?.emailConfirmed).toBe(false);
   });
 
-  it('should return 400 for invalid email format', () => {
+  it('should return 400 for invalid email format', async () => {
```
Comment (savitasawanteisg-byte) [inline src/__tests__/api/register.test.ts:96]: **Code quality: Excessive as any type casts**

await handleRegister(req as any, res as any) defeats the purpose of TypeScript. The mock objects should properly implement the ApiRequest and ApiResponse interfaces without needing as any.

**Review by savitasawanteisg-byte — CHANGES_REQUESTED** [review]
## Code Review - PR #343: Input Sanitization Middleware

Overall the approach is sound. Using the xss library with an empty whitelist + stripIgnoreTagBody is a good strategy. However, there are several security gaps, a potential DoS vector, and code quality issues that need to be addressed before merging.

**Summary of Issues:**
1. **Security (High):** No sanitization of req.query or req.params
2. **Security (Medium):** Prototype pollution risk via __proto__ key
3. **Reliability (Medium):** No recursion depth limit (DoS)
4. **Testing:** Missing edge-case tests
5. **Code quality:** Missing newline at EOF, excessive as any casts

Please address these issues and re-submit for review.

**Review by savitasawanteisg-byte — DISMISSED** [review]
## Code Review — PR #343: fix(security): sanitize request bodies against XSS #247

PR reviewed and approved for merge. Review completed.

**✅ All checks pass. All issues from the previous review have been addressed.**

**What was fixed since last review:**
- ✅ `MAX_DEPTH = 20` recursion guard added — DoS via stack overflow prevented
- ✅ `DANGEROUS_KEYS` set blocks `__proto__`, `constructor`, `prototype` — prototype pollution prevented
- ✅ `req.query` and `req.params` now also sanitized — complete XSS coverage
- ✅ Function renamed to `sanitizeRequest` with backward-compatible `sanitizeRequestBody` alias
- ✅ Object traversal uses `for...of` loop with early `DANGEROUS_KEYS` check — cleaner than `.map()` approach

**Minor non-blocking notes:**
- `index.ts` imports the old alias `sanitizeRequestBody` — consider updating to `sanitizeRequest` for consistency since the alias exists only for backward compatibility
- The route comment still says 'body sanitization' — should read 'request sanitization' since params/query are now covered too
- Missing newline at EOF in `sanitize-input.ts`

Well done addressing all the security feedback. This is a solid implementation.
